File: R/packages/euler/R/Problem03.R
# Author: Aaron Johnson
# Date: 2016-09-04

# Largest prime factor
# The prime factors of 13195 are 5, 7, 13 and 29.
# What is the largest prime factor of the number 600851475143 ?

library(hash)

primes <- function() {
	# wrap hash library functions - only key strings are supported
	haskey <- function(k, m) has.key(as.character(k), m)
	setkey <- function(k, v, m) m[[as.character(k)]] <- v
	getkey <- function(k, m) m[[as.character(k)]]
	delkey <- function(k, m) delete(as.character(k), m)

	# sieve is passed by reference
	addNextComposite <- function(n, prime, sieve) {
		while (haskey(n, sieve)) n <- n + prime
		setkey(n, prime, sieve)
	}

	primeGen <- function(n, sieve) {
		# mimic tail-call recursion optimization with a while-loop
		while (haskey(n, sieve)) {
			prime <- getkey(n, sieve)

			delkey(n, sieve)
			addNextComposite(n + prime, prime, sieve)
			n <- n + 1
		}

		setkey(n*n, n, sieve)
		list(getnext=function() primeGen(n + 1, sieve),
			 current=n)
	}

	primeGen(2, hash())
}

primeFactors <- function(x) {
	pfacts <- c()
	quotient <- x
	remainder <- 0
	prime <- primes()
	p <- prime$current

	while (p <= quotient) {
		remainder <- quotient %% p
		while (remainder == 0) {
			quotient <- quotient %/% p
			remainder <- quotient %% p
			pfacts <- append(pfacts, p)
		}

		prime <- prime$getnext()
		p <- prime$current
	}

	pfacts
}

problem03 <- function() {
	max(primeFactors(600851475143))
}

File: euler/hs/problems/src/Problem03.hs
-- Author: Aaron Johnson
-- Date: 2016-09-05

{-
Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?
-}

module Problem03 (solve, primes, primeFactors) where
import qualified Data.Map as Map

primes :: [Int]
primes =
    prime_ 2 Map.empty
    where
        addNextComposite :: Int -> Int -> Map.Map Int Int -> Map.Map Int Int
        addNextComposite n prime sieve
            | Map.member n sieve = addNextComposite (n + prime) prime sieve
            | otherwise          = Map.insert n prime sieve
        prime_ :: Int -> Map.Map Int Int -> [Int]
        prime_ n sieve = case Map.lookup n sieve of
            Just prime ->
                let gc = Map.delete n sieve
                    ac = addNextComposite (n + prime) prime gc
                in prime_ (n + 1) ac
            Nothing    ->
                let ac = Map.insert (n*n) n sieve
                in n : prime_ (n + 1) ac

primeFactors :: Int -> [Int]
primeFactors n =
    pfactor primes n
    where
        pfactor :: [Int] -> Int -> [Int]
        pfactor [] _ = error "Empty prime list"
        pfactor ps@(prime:pt) n
            | prime >= n         = [prime]
            | n `mod` prime == 0 = prime : pfactor ps (n `quot` prime)
            | otherwise          = pfactor pt n

solve :: Integer
solve = toInteger (maximum $ primeFactors 600851475143)